test/e2e/framework/timeouts.go

```diff
@@ -89,26 +97,31 @@ type TimeoutContext struct {
 
 	// SnapshotControllerMetrics is how long to wait for snapshot controller metrics.
 	SnapshotControllerMetrics time.Duration
+
+	// SystemPodsStartup is how long to wait for system pods to be running.
+	SystemPodsStartup time.Duration
+
+	// NodeSchedulable is how long to wait for all nodes to be schedulable.
+	NodeSchedulable time.Duration
+
+	// SystemDaemonsetStartup is how long to wait for all system daemonsets to be ready.
+	SystemDaemonsetStartup time.Duration
 }
 
 // NewTimeoutContextWithDefaults returns a TimeoutContext with default values.
+// Some of these timeouts are configurable, therefore
+// NewTimeoutContextWithDefaults should be called after command line parsing.
 func NewTimeoutContextWithDefaults() *TimeoutContext {
-	return &TimeoutContext{
-		PodStart:                  podStartTimeout,
-		PodStartShort:             podStartShortTimeout,
-		PodStartSlow:              podStartSlowTimeout,
-		PodDelete:                 podDeleteTimeout,
-		ClaimProvision:            claimProvisionTimeout,
-		ClaimProvisionShort:       claimProvisionShortTimeout,
-		DataSourceProvision:       dataSourceProvisionTimeout,
-		ClaimBound:                claimBoundTimeout,
-		PVReclaim:                 pvReclaimTimeout,
-		PVBound:                   pvBoundTimeout,
-		PVCreate:                  pvCreateTimeout,
-		PVDelete:                  pvDeleteTimeout,
-		PVDeleteSlow:              pvDeleteSlowTimeout,
-		SnapshotCreate:            snapshotCreateTimeout,
-		SnapshotDelete:            snapshotDeleteTimeout,
-		SnapshotControllerMetrics: snapshotControllerMetricsTimeout,
-	}
+	// Make a copy, otherwise the caller would have the ability to
+	// modify the defaults.
+	copy := TestContext.timeouts
```

Comment:
but this no longer have the defaults, does it?

are not the flags overrding the defaults?

Reply:
the defaults are `copy := defaultTimeouts`

Reply:
What I meant is that we should not return `&TestContext.timeouts` - that would give the caller read-write access. Let me change the comment to:
> Make a copy, otherwise the caller would have the ability to modify the values.

With "defaults" I was following the function name ("WithDefaults"). That "the defaults" are configurable gets ignored here. It's debatable whether that function name is a good name - probably not. But changing it doesn't seem worthwhile.

Reply:
I checked and renaming the function isn't affecting that much code - let's do it.

Reply:
I'm confused, let me see if I'm reading this wrong:
Before: `NewTimeoutContextWithDefaults` always returned the timeouts constants defined here, they had to use `NewFrameworkWithCustomTimeouts` to use the custom timeouts
Now:  they CustomTimeouts are always used

Reply:
Before: `NewTimeoutContextWithDefaults` returned a subset of the timeouts provided by the framework, with defaults that were hard-coded in the source code of the framework.

After: `NewTimeoutContext` returns all timeouts provided by the framework, of which some are configurable via command line parameters. We still need to decide whether we want to make all of them configurable, and how (more command line parameters or config file?).

The "all timeouts" may be a bit too optimistic, though, but it is the goal. There are several hard-coded timeouts (for example, in e2e/framework/pod/wait.go) that either should be replaced by the ones in TimeoutContext (most likely) or need to be added there (less likely). This needs to be checked on a case-by-case basis after we agree on the general concept.

That the new timeout context then gets passed back to the framework by the CSI driver setup code without changes (as far as I can tell) is a bit odd, but I guess it was added because some CSI driver might want to run with different timeouts.

Reply:
I missed that test/e2e/storage/drivers/in_tree.go changes overwrites some timeouts - so that probably explains this aspect of the CSI driver setup.

Reply:
clarified offline, only the Timeouts passed in the flags are modified